### ISSUE TITLE
Simplify implementation of back-pressure in generated stubs

### ISSUE THREAD

**Issue by jhump** [grpc/grpc-java#1549]
Pending API changes can allow reactive/async pattern for interacting with flow control and applying back pressure: https://github.com/grpc/grpc-java/pull/1545/files

In many cases, automatic back-pressure in generated stubs could be very useful -- e.g. having calls to `StreamObserve#onNext(T)` block instead of queueing.

It's been pointed out that this could cause deadlock for bidi-streaming operations, so perhaps we can just not expose this functionality for bidi-streaming calls?

It may also be worth pointing out that most other runtimes (wrapped languages and Go) already expose streams via blocking operations and already require that apps be aware of and work-around possible deadlock issues resulting therefrom. So maybe providing similar mechanisms in Java is fine, with said caveats.

Another possible alternative could possibly be done in an extension/add-on instead of in GRPC. For example, wrapping streaming requests and responses with RxJava `Observables` may further simplify the async case enough to make the synchronous (and possibly-deadlock-prone) case unnecessary.

**Comment by ejona86** [comment]
> It may also be worth pointing out that most other runtimes (wrapped languages and Go) already expose streams via blocking operations and already require that apps be aware of and work-around possible deadlock issues resulting therefrom. So maybe providing similar mechanisms in Java is fine, with said caveats.

This could possibly be as an alternative to StreamObserver. For example, we could make a more robust blocking API that could handle bidi. To my knowledge, the various languages' APIs are either completely blocking or completely async; they don't "randomly" block.

But alternatively, simply having the blocking behavior be opt-in could be enough to address deadlock concerns.

**Comment by ejona86** [comment]
The StreamObserver used for sending messages currently completely ignores outbound flow control. The original argument for this is that most applications don't care. While not debating that point, it does seem enough applications do care that they experience a rude awakening.

In discussing with @louiscryan, it seems we are agreeing that it may have been a misstep to ignore flow control by default. If doing it again, we may have had blocking for sending enabled by default. However, changing the behavior now would impact existing applications and it isn't clear how much pain that would cause our existing users. So it may be prudent to introduce something new that we encourage users to migrate to.

In any case, we do still feel the need for a non-blocking variant, where the application should use `isReady()`/`onReady()` (which is not yet available on the async client stub).

In a world where some StreamObservers block when sending and others do not, we need to make it easy for our users to make sure they are using the correct one. So although it may be easy for us to have blocking by default and then an option on the returned StreamObserver to use non-blocking, this seems prone to error. Similarly, having non-blocking by default and a utility to convert it to blocking seems prone to error. Having to audit every call site to make sure a method call is present to get the preferred behavior seems untenable. So instead, we may want a different stub creation method, although this could have complications for injection since they type does not fully define the behavior.

This discussion is also impacted by the client-side stub not having an easy way to add `onReady()` notification due to races. A worsening of an existing race is an implementation of `onReady()` would need a reference to the outbound StreamObserver, but that is only available after the call was started. But the opposite is also a problem: if `onReady()` is configured after the call is started then notifications may be missed. Attempts to save missing notifications gets complex because the executor to use isn't available and it is important not to call onReady() after onClose().

Our answer for client-side is still use use `Channel/ClientCall` directly, but the lack of a generated stub, being forced to use a different API, and the need to call `request()` contribute to users seeming to be disappointed by the solution.

Using RxJava's Observables (or similar) would simplify the flow control aspect of this discussion, and would allow reversing the input/outputs (#1475) to be more obvious. However, it also causes a dependency problem for our stub generation, since users using the blocking stub should not need to depend on RxJava. We'd also want to ensure that we'd be able to add additional gRPC-specific methods over time.

One important but side point, is that flow control currently includes push-back related to RPC creation. `onReady()` is only called after the call has be accepted by a transport (transport must be connected and not exceed MAX_CONCURRENT_STREAMS), so even unary RPCs would be impacted. Granted, it is good that we are observing flow control in more cases, but it is worrisome that unary RPCs would be impacted since they are so much more prevalent. It's unclear whether we would want to make tweaks for the unary use-case.

**Comment by ejona86** [comment]
We discovered a fundamental issue yesterday on doing blocking from the async interface. If you send from a callback, then you will block until onReady is called. But onReady will be queued behind the currently running callback (the one you are blocking within) and never run. Oops.

```java
requestObserver = asyncStub.recordRoute(responseObserver);
StreamObserver<Response> responseObserver = new StreamObserver<Response>() {
  @Override
  public void onNext(Response resp) {
    requestObserver.onNext(req); // Blocks onReady callback!
  }
  ...
};
```

It is possible to fix, but is ugly enough (executors, executors everywhere) that it doesn't seem like a great idea. So we're trying to figure out what to do now.

**Comment by jhump** [comment]
I'm not sure I follow. I'm guessing this is because of how the library serializes calls to a given listener. But it's not clear why these two operations would be serialized together -- I would have thought that they'd be independent. Or is this a concern with using a parallelism-constrained `Executor`, which could deadlock due not creating an additional thread to run `onReady`?

**Comment by ejona86** [comment]
@jhump, all callbacks are serialized. That is pretty necessary for onMessage and onClose. But we serialize all callbacks for a listener to avoid implementations of the listener from needing to worry about thread-safety (because they'd get it wrong).

**Comment by jhump** [comment]
Yeah, I see now. I think I just misunderstood your example, so the crux of the problem didn't jump out at me.

**Comment by louiscryan** [comment]
This PR https://github.com/grpc/grpc-java/pull/1842 is somewhat relevant here

**Comment by louiscryan** [comment]
... in particular we can now layer a 'blocking' stream observer over the underlying stream observer returned by the runtime

**Comment by hsaliak** [comment]
Removing this from 1.0 /cc @louiscryan

**Comment by cbornet** [comment]
Is there any way to apply back-pressure today ? There seem to be some utility functions but I have no clue how to use them. Can someone point to docs or examples ?
Also, to prevent deadlocks, it would be nice to use the reactive stream pattern through rxjava2's Flowable.

**Comment by carl-mastrangelo** [comment]
I don't think there is documentation today, but the intended way is to cast the StreamObserver to a CallStreamObserver, disable automatic flow control, and call `request` and `isReady` appropriately.

**Comment by stephenh** [comment]
@cbornet FWIW this is what I came up with ~several months ago (poke around the rest of the project for the callers/usage):

https://github.com/stephenh/mirror/blob/master/src/main/java/mirror/BlockingStreamObserver.java

Disclaimer this is my hobby project and it "works great" AFAICT, but I make no claims to intimately understanding grpc-java internals, so YMMV. Just passing along the link in case it saves you some time.

**Comment by cbornet** [comment]
@stephenh thanks a lot. That will be very helpful. My ultimate goal would be to have non-blocking back-pressure but that's certainly a good step towards that direction.

**Comment by carl-mastrangelo** [comment]
@stephenh that is going to be really easy to consume all your executor threads.  Also, it will mysteriously and silently drop messages if an IE comes in.

**Comment by cbornet** [comment]
@carl-mastrangelo I'm trying to understand if the `onReady/request` can bring end-to-end back pressure from a client to a server (or vice-versa). Is it the case ? I mean : is there some flow control at the transport level so that the client receives `onReady` only when the server calls `request` or does `onReady` only means that the message was sent to the server and buffered in some incoming queue of the server ?

Also, it would be great to have a grpc-rx library that would bring ready-to-use adapters for people wanting to use rxJava or Reactor.

**Comment by stephenh** [comment]
@carl-mastrangelo I'm very likely being naive here, but do you mean my application's own threads, or the internal grpc threads?

My intention was specifically to block my application thread(s), because I'm sending N files out over grpc-java, and I don't want to read in the bytes for all N files and have them queue them up in the grpc queue, as then it will OOME. So I want my app thread that calls `onNext(messageWithFileBytesInIt)` to block.

If this `BlockingStreamObserver` will cause some internal grpc IO threads to block, then yes, that's unintended and would appreciate pointers on a better approach.

Per dropping messages on IE, yes, that's true, I should probably at least log about that, but AFAIU / for my app, an IE would only happen when the JVM/app is shutting down, e.g. on ctrl-c, so dropping a message there/for my app, is fine/desirable.

**Comment by cbornet** [comment]
After some reading, it does seem there is flow control on the wire at the HTTP2 level. So it should be possible to build a reactive streams / back-pressured implementation. The main problem is the lack of documentation and examples which make it very hard for non gRPC experts like me to output something working.

**Comment by carl-mastrangelo** [comment]
@stephenh Yes, you app threads, as provided to `executor()` when you set up your Channel / Server.  If you stall all your threads, none of them will be able to take more traffic.  The pushback logic probably needs to live outside gRPC code, and not inline as in your example above.  If you are calling the stubs from your own threads (not on the executor you gave to the channel) then you are probably fine blocking.  

IE is actually a "collaborative" feature.  You have to want an IE in the reciever, and also have to intentionally interrupt another thread (such as on `Future.cancel(true)`).  The trouble comes from code you may not own interrupting your code.  If you aren't expecting an IE, its better to just let it propagate.  


@cbornet There are two levels of FC: wire and API.  The gRPC API is message based, which is converted to byte based in the library.  I agree, this is not well covered anywhere.  One effect that is not obvious is that unary calls generally don't honor the message flow control, but do honor the byte flow control.  That means it is still easy to overload your sender if you don't pay attention to `isReady`, because they will buffer up in memory.  One day I will write a comprehensive guide on how this works, but that's extremely time consuming and thus hasn't been a priority.

**Comment by stephenh** [comment]
> If you are calling the stubs from your own threads (not on the executor you gave to the
> channel) then you are probably fine blocking.

Yep, that's what I'm doing. Thanks!

**Comment by bobymicroby** [comment]
@carl-mastrangelo Hi Carl. Since writing proper docs can be time consuming, can you make a simplified gist showing how to make bidi streaming with flow control using reactive streams API for example ?

**Comment by bobymicroby** [comment]
I've seen people already trying to do it, but having hard time : https://github.com/xiaodongw/grpc-rx/issues/1

**Comment by cbornet** [comment]
See #3119 for a [flow control example](https://github.com/grpc/grpc-java/tree/master/examples/src/main/java/io/grpc/examples/manualflowcontrol)
and https://github.com/salesforce/grpc-java-contrib/tree/master/rxgrpc for RxJava2 generator plugin with flow control enabled.

**Comment by davidraleigh** [comment]
@cbornet updated link is https://github.com/salesforce/reactive-grpc, correct?

**Comment by cbornet** [comment]
@davidraleigh exactly ! You can now generate RxJava 2 and Reactor bindings.

**Comment by ulfjack** [comment]
Any updates here? We have a server implementation and we need to apply flow control. We have a server thread that's generating data, but we need to block it when the client is accepting data too slowly to avoid OOMing (our use case is not a typical network server, and this is fine for us).

In normal use, I can see cases where isReady returns false, and I'm using a similar approach to @stephenh to block. However, in an inprocess test server, I never see isReady returning false; instead, onNext appears to block. That makes it impossible to test the code using the inprocess test server.

I'm also concerned about race conditions where the server thread checks isReady and then goes to sleep, but the callback comes in between the isReady call and actually going to sleep. I think that can't happen if both synchronize on the same external object, but the API does not actually specify how isReady and onReady are internally synchronized.

**Comment by OlegDokuka** [comment]
@ulfjack. Please, consider using RSocket which supports backpressure correctly - http://rsocket.io/

Cheers.

**Comment by ejona86** [comment]
> In normal use, I can see cases where isReady returns false, and I'm using a similar approach to @stephenh to block. However, in an inprocess test server, I never see isReady returning false; instead, onNext appears to block. That makes it impossible to test the code using the inprocess test server.

@ulfjack, if you are using `directExecutor()` then the client and server share a single thread, which makes the tests deterministic. Simply remove at least one of the calls that specify `directExecutor()` and `onNext()` will then be processed asynchronously.

> I'm also concerned about race conditions where the server thread checks isReady and then goes to sleep, but the callback comes in between the isReady call and actually going to sleep. I think that can't happen if both synchronize on the same external object

The race totally seems possible. I don't see how any locking in gRPC could prevent it; if you added a `sleep(1 minute)` between the two parts, it seems obviously racy.

> the API does not actually specify how isReady and onReady are internally synchronized.

The only guarantee is that if `isReady()` returns `false` (there is no guarantee that your application has observed it yet) there will be an `onReady()` callback at some point when `isReady() == true`. Basically, "no need to poll; we'll tell you when it changes."

Note that it does not imply the converse: "spurious" `onReady()` callbacks are possible, so it is possible for `isReady() == false` within the `onReady()` callback. This is due to races between gRPC delivering a callback and the application writing more data. (So it _was_ ready again, but it became non-ready by the time onReady() was called.)

Hmmm... It seems that isn't well-specified in the documentation.

**Comment by ulfjack** [comment]
@ejona86 Good catch! I managed to write some tests using the inprocess test server.

The code posted by @stephenh does not pass my tests. In particular, if the client cancels the call while a thread is in onNext, the onNext hangs. The OnCancel handler correctly wakes up the thread in onNext, but that thread only checks isReady(), which remains false, so the loop does not terminate.

**Comment by h4yin** [comment]
> @stephenh Yes, you app threads, as provided to `executor()` when you set up your Channel / Server. If you stall all your threads, none of them will be able to take more traffic. The pushback logic probably needs to live outside gRPC code, and not inline as in your example above. If you are calling the stubs from your own threads (not on the executor you gave to the channel) then you are probably fine blocking.
> 
> IE is actually a "collaborative" feature. You have to want an IE in the reciever, and also have to intentionally interrupt another thread (such as on `Future.cancel(true)`). The trouble comes from code you may not own interrupting your code. If you aren't expecting an IE, its better to just let it propagate.
> 
> @cbornet There are two levels of FC: wire and API. The gRPC API is message based, which is converted to byte based in the library. I agree, this is not well covered anywhere. One effect that is not obvious is that unary calls generally don't honor the message flow control, but do honor the byte flow control. That means it is still easy to overload your sender if you don't pay attention to `isReady`, because they will buffer up in memory. One day I will write a comprehensive guide on how this works, but that's extremely time consuming and thus hasn't been a priority.

Wish your guide, it would be very helpful.

**Comment by phamtai97** [comment]
I implemented model **back-perssure**, but I have the following problem: before I write message the stream, I check "isCanceled()" and it returns true. I see the document that "isCanceled" is true when the peer closed the stream. So, I check the client and it does not close the stream.

So, I think that the server or the client is set **keepAliveTimeout** which is a small value. I increased the value of this config, but the **isCanceled()" is still true.

How should I handle it to avoid making such a mistake?
Thank all.
@ejona86

**Comment by ejona86** [comment]
@phamtai97, your issue sounds different than what is here. Please open a new issue to discuss.

**Comment by phamtai97** [comment]
@ejona86 oh, I opened this issues in here: https://github.com/grpc/grpc-java/issues/7984
